### PR TITLE
Update SSH to UART example for ESP-IDF v5.x, in-place cmake

### PR DESCRIPTION
**Pull request by gojimmypi**
This PR updates the ESP32 SSH to UART example to work with the Espressif ESP-IDF v5.x environment, as well as use the updated `CMakeLists.txt` file that allows for in-place compiling without needing to copy wolfssl and/or wolfssh locally to the project.

Note that for best results, https://github.com/wolfSSL/wolfssl/pull/7173 and https://github.com/wolfSSL/wolfssh/pull/649 should be merged first.

There's a known problem with SHA256 HW Acceleration for this project, so it has been disabled in the local `components/wolfssl/user_settings.h` with `NO_WOLFSSL_ESP32_CRYPT_HASH_SHA256`. A solution for that is still WIP.

See enclosed `README.md` files for more details.

## Edit:

Note the wolfSSL and wolfSSH libraries need to be findable by the cmake file. If everything is cloned from the same directory, it should automatically just work. 

The cmake file will look for `wolfssh-username` and `wolfssl-username` directories, and then look for `wolfssh` and `wolfssl` directories when the `WOLFSSL_ROOT` and/or `WOLFSSH_ROOT` environment variables are not defined. 

For example, if everything is cloned from the `C:\workspace` directory:

```text
c:\workspace\
c:\workspace\wolfssl
c:\workspace\wolfssh
c:\workspace\wolfssh-examples
```

The project at `c:\workspace\wolfssh-examples\Espressif\ESP32\ESP32-SSH-Server` should be able to find wolfSSL and wolfSSH libraries in the above structure.

If instead the `wolfssh-examples` is cloned to `c:\test\wolfssh-examples`, then the user _must_ define the locations of wolfSSL and wolfSSH libraries manually.

The directories can be defined as described in the [README.md](https://github.com/gojimmypi/wolfssh-examples/blob/PR-update-Espressif/Espressif/ESP32/ESP32-SSH-Server/README.md#linux-quick-start) in the project [CMakeLists.txt](https://github.com/gojimmypi/wolfssh-examples/blob/d14f9b1b07290b010af3d1c60db836e7f5c2367e/Espressif/ESP32/ESP32-SSH-Server/CMakeLists.txt#L28)

At some point once the codebase is all stabilized, we may consider using [Managed Components](https://components.espressif.com/components?q=namespace:wolfssl) that would allow for a "clone & run" with relatively little fuss.

See https://github.com/wolfSSL/wolfssl/issues/6234 for a roadmap of Espressif updates.

### CODE REVIEW

**Review by ejohnstown — DISMISSED** [review]
In general I'm fine with this. A couple things:

1. Copyright years. This is something that will be updated next release.
2. One or two files had comments with their file path. They weren't updated from when they used to be wolfSSL files.
3. Several files are missing an LF on the last line. I'm guessing this is a VSCode thing.
4. Several files have C++ style comments instead of only C style. Again, probably a VSCode thing.

If this runs fine for Hide-san for his needs, this can be merged. While I marked Approve, it is conditional on Hide.

File: Espressif/ESP32/ESP32-SSH-Server/components/wolfssl/include/user_settings.h [hunk 1]
```diff
@@ -0,0 +1,589 @@
...
+
+#undef WOLFSSL_ESPIDF
+#undef WOLFSSL_ESP32
+#undef WOLFSSL_ESPWROOM32SE
+#undef WOLFSSL_ESP32
+#undef WOLFSSL_ESP8266
+
+#define WOLFSSL_ESPIDF
+
+#define DEBUG_WOLFSSH
```
Comment (miyazakh) [inline Espressif/ESP32/ESP32-SSH-Server/components/wolfssl/include/user_settings.h:32]: can we disable this as default?

File: Espressif/ESP32/ESP32-SSH-Server/components/wolfssl/include/user_settings.h [hunk 2]
```diff
@@ -0,0 +1,589 @@
...
+#ifdef ESP_ENABLE_WOLFSSH
+    /* The default SSH Windows size is massive for an embedded target. Limit it: */
+    #define DEFAULT_WINDOW_SZ 2000
+
+    /* These may be defined in cmake for other examples: */
+    #undef  WOLFSSH_TERM
+    #define WOLFSSH_TERM
+
+    #undef  DEBUG_WOLFSSH
+    #define DEBUG_WOLFSSH
```
Comment (miyazakh) [inline Espressif/ESP32/ESP32-SSH-Server/components/wolfssl/include/user_settings.h:56]: can we disable this as default?

File: Espressif/ESP32/ESP32-SSH-Server/main/main.c [hunk 3]
```diff
@@ -242,42 +130,65 @@ bool NoEthernet()
...
+    }
     /* TODO ShowCiphers(); */
-#endif
+    #endif
 
+    /* Set time for cert validation.
+     * Some lwIP APIs, including SNTP functions, are not thread safe. */
+    ret = set_time(); /* need to setup NTP before WiFi */
+
+#ifndef DISABLE_SSH_UART
```
Comment (miyazakh) [inline Espressif/ESP32/ESP32-SSH-Server/main/main.c:188]: Is it possible to just send message to SSH client when disabled UART instead to put messages to UART? This should be nice if user doesn't have UART connection.

**Review by miyazakh — CHANGES_REQUESTED** [review]

File: Espressif/ESP32/ESP32-SSH-Server/main/main.c (hunk 3 above)
Comment (gojimmypi) [reply Espressif/ESP32/ESP32-SSH-Server/main/main.c:188]: @miyazakh I can certainly make the default to be no wolfSSH / wolfSSL debugging messages. I had purposely left it on as it seemed more interesting for an example. 

Would you prefer I also turn off other verbosity, such as when characters are sent / received? For example, I've disabled all wolfSSL and wolfSSH debugging, yet these messages still appear in the serial port at the server:

```text
I (99788) ssh_server: Received 1 bytes from client.
I (99795) uart_helper: UART Send Data
I (99795) TX_TASK: Wrote 1 bytes
I (124278) SSH Server main: wolfSSH Server main loop heartbeat!
```

There's also a variety of other initialization not controlled by the wolfSSL gating macros. I could make it more silent with instructions for verbosity in the `README.md`. 

Let me know which you prefer.

Comment (gojimmypi) [reply Espressif/ESP32/ESP32-SSH-Server/main/main.c:188]: That's an interesting request. The whole point of the example is the SSH to UART, but let me see if I can do something for only the SSH server part with no connected UART.

How did you envision this working? Perhaps just echo the characters to the server console USB connected UART, like a debug message, noted [above](https://github.com/wolfSSL/wolfssh-examples/pull/4#discussion_r1479125590) in debugging?

**Comment by gojimmypi** [comment]
> In general I'm fine with this. A couple things: [Copyright / License / LF / Comment Style]

@ejohnstown  I've pushed an update to address these items. I'll squash when ready to merge, if desired.

File: Espressif/ESP32/ESP32-SSH-Server/main/main.c (hunk 3 above)
Comment (miyazakh) [reply Espressif/ESP32/ESP32-SSH-Server/main/main.c:188]: I mean that server sends message to the ssh client terminal, like welcome message does.
Teraterm

Comment (miyazakh) [reply Espressif/ESP32/ESP32-SSH-Server/main/main.c:188]: It appears that setting `SSH_SERVER_ECHO` could achieves the purpose.

**Review by miyazakh — APPROVED** [review]

**Review by ejohnstown — APPROVED** [review]

**Pull request merged**